content/blog/2025-04-27-valkey-modules-rust-sdk-updates.md

```diff
@@ -0,0 +1,276 @@
+    }
+}
+#[shutdown_event_handler]
+fn shutdown_event_handler(_ctx: &Context, _event: u64) {
+    ...
+}
+#[key_event_handler]
+fn key_event_handler(ctx: &Context, key_event: KeyChangeSubevent) {
+    match key_event {
+        KeyChangeSubevent::Deleted => {}
+        KeyChangeSubevent::Evicted => {}
+        KeyChangeSubevent::Overwritten => {}
+        KeyChangeSubevent::Expired => {}
+    }
+}
+```
+
+## Custom ACL categories support
+
+Valkey 8 introduced support for custom ACL categories which simplifies access control for custom commands introduced in a module.
+To implement that we need to enable `required-features = ["min-valkey-compatibility-version-8-0"]` in `Cargo.toml` and register new categories in `valkey_module!` macro.
+Then we can restrict our custom commands to these custom ACL categories.
+
+```rust
+valkey_module! {
+    ...
+    acl_categories: [
+        "acl_one",
+        "acl_two"
+    ]
+    commands: [
+        ["cmd1", cmd1, "write",  0, 0, 0, "acl_one"],
+        ["cmd2", cmd2, "", 0, 0, 0, "acl_one acl_two"],
+    ]
+```
+
+## Validating / Rejecting CONFIG SET
+
+Configuration flexibility is important but so is validation.
+The SDK now supports specifying optional callback functions to validate or reject custom configurations.
+This is available for `String`, `i64`, `bool` and `enum` configs.  
+Here is an example of such validation for i64 custom config.  
+
+```rust
+lazy_static! {
+    static ref CONFIG_I64: ValkeyGILGuard<i64> = ValkeyGILGuard::default();
+}
+fn on_i64_config_set<G, T: ConfigurationValue<i64>>(
+    config_ctx: &ConfigurationContext,
+    _name: &str,
+    val: &'static T,
+) -> Result<(), ValkeyError> {
+    if val.get(config_ctx) == custom_logic_here {
+        log_notice("log message here");
+        Err(ValkeyError::Str("error message here "))
+    } else {
+        Ok(())
+    }
+}
+valkey_module! {
+    configurations: [
+        i64: [
+            ["my_i64", &*CONFIG_I64, 10, 0, 1000, ConfigurationFlags::DEFAULT, Some(Box::new(on_configuration_changed)), Some(Box::new(on_i64_config_set::<ValkeyString, ValkeyGILGuard<i64>>))],
+        ],
+        ...
+    ]
+}
+```
+
+## Defrag
+
+For memory-sensitive applications, defragmentation is essential.
+The SDK now offers a safe and idiomatic Rust abstraction over the defrag API for custom data types.
+The new `Defrag` struct abstracts away the raw C FFI calls.
+
+```rust
+static MY_VALKEY_TYPE: ValkeyType = ValkeyType::new(
+    "mytype123",
+    0,
+    raw::RedisModuleTypeMethods {
+        ...
+        defrag: Some(defrag),
+    },
+);
+unsafe extern "C" fn defrag(
+    defrag_ctx: *mut raw::RedisModuleDefragCtx,
+    _from_key: *mut RedisModuleString,
+    value: *mut *mut c_void,
+) -> i32 {
+    let defrag = Defrag::new(defrag_ctx);
+    ...
+    0
+}
+valkey_module! {
+    ...
+    data_types: [
+        MY_VALKEY_TYPE,
+    ],
+    ...
+}
+```
+
+## Redis support
+
+Need your module to work with both Valkey and recent versions Redis?
```

Comment:
missing word: recent versions **of** Redis

Reply:
updated